File: crafting-interpreters/python/lox/interpreter.py
from .environment import Environment
from .token_types import TokenType
from .error_logger import RuntimeException, ErrorLogger


OPERATOR_MAP = {
    TokenType.LESSER: "__lt__",
    TokenType.LESSER_EQUAL: "__le__",
    TokenType.GREATER: "__gt__",
    TokenType.GREATER_EQUAL: "__ge__",
    TokenType.STAR: "__mul__",
    TokenType.SLASH: "__div__",
    TokenType.MINUS: "__sub__",
    TokenType.EQUAL_EQUAL: "__eq__",
}


class Interpreter:
    def __init__(self):
        self.env = Environment()

    def interpret(self, statements):
        try:
            for stmt in statements:
                self.execute(stmt)
        except RuntimeException as ex:
            ErrorLogger.runtime_error(ex)

    def evaluate(self, expr):
        """
        Evaluate an expression
        """
        return expr.accept(self)

    def execute(self, stmt):
        """
        Execute a statement
        """
        stmt.accept(self)

    def isTruthy(self, expr):
        """
        In Lox, None is Falsy
        rest all are Truthy
        """
        if expr is None:
            return False
        if isinstance(expr, bool):
            return expr
        return True

    def checkNumberOperands(self, token, *values):
        for value in values:
            if isinstance(value, (int, float)):
                continue
            raise RuntimeException(token, " operand/s must be number/s")

    def visitLiteralExpr(self, expr):
        return expr.value

    def visitGroupingExpr(self, expr):
        return self.evaluate(expr.exp)

    def visitUnaryExpr(self, expr):
        val = self.evaluate(expr.exp)

        if expr.op.token_type == TokenType.MINUS:
            return -val
        if expr.op.token_type == TokenType.NOT:
            return not val

    def visitBinaryExpr(self, expr):
        left = self.evaluate(expr.left)
        right = self.evaluate(expr.right)

        tk_type = expr.op.token_type

        if tk_type == TokenType.PLUS:
            if isinstance(left, (int, float)) and isinstance(right, (int, float)):
                return left + right
            if isinstance(left, str) and isinstance(right, str):
                return left + right

            raise RuntimeException(expr.op, "operands must be numbers or strings")

        self.checkNumberOperands(expr.op, left, right)
        op_name = OPERATOR_MAP[tk_type]
        fn = getattr(left, op_name)
        return fn(right)

    def visitExpressionStmt(self, stmt):
        self.evaluate(stmt.exp)

    def visitPrintStmt(self, stmt):
        val = self.evaluate(stmt.exp)
        print(val)

    def visitVarStmt(self, stmt):
        val = self.evaluate(stmt.value)
        name = stmt.name.lexeme
        self.env.define(name, val)

    def visitVariableExpr(self, stmt):
        name = stmt.name
        return self.env.get(name)

File: crafting-interpreters/java/lox/Interpreter.java
package lox;
import java.util.List;
import static lox.TokenType.*;

class Interpreter implements Expr.Visitor<Object>, Stmt.Visitor<Void> {
  private Environment environment = new Environment();

  private Object evaluate(Expr expr) { return expr.accept(this); }
  private void execute(Stmt stmt) { stmt.accept(this); }

  void executeBlock(List<Stmt> statements, Environment environment) {
    Environment previous = this.environment;
    try {
      this.environment = environment;

      for (Stmt statement : statements) {
        execute(statement);
      }
    } finally {
      this.environment = previous;
    }
  }

  private boolean isTruthy(Object object) {
    if (object == null)
      return false;
    if (object instanceof Boolean)
      return (boolean)object;
    return true;
  }
  private boolean isEqual(Object a, Object b) {
    // nil is only equal to nil.
    if (a == null)
      return b == null;

    return a.equals(b);
  }
  private void checkNumberOperand(Token operator, Object operand) {
    if (operand instanceof Double)
      return;
    throw new RuntimeError(operator, "Operand must be a number.");
  }
  private void checkNumberOperands(Token operator, Object left, Object right) {
    if (left instanceof Double && right instanceof Double)
      return;

    throw new RuntimeError(operator, "Operands must be numbers.");
  }
  private String stringify(Object object) {
    if (object == null)
      return "nil";

    // Hack. Work around Java adding ".0" to integer-valued doubles.
    if (object instanceof Double) {
      String text = object.toString();
      if (text.endsWith(".0")) {
        text = text.substring(0, text.length() - 2);
      }
      return text;
    }

    return object.toString();
  }
  @Override
  public Object visitLiteralExpr(Expr.Literal expr) {
    return expr.value;
  }

  @Override
  public Object visitGroupingExpr(Expr.Grouping expr) {
    return evaluate(expr.expression);
  }

  @Override
  public Object visitUnaryExpr(Expr.Unary expr) {
    Object right = evaluate(expr.right);

    switch (expr.operator.type) {
    case BANG:
      return !isTruthy(right);
    case MINUS:
      checkNumberOperand(expr.operator, right);
      return -(double)right;
    }

    // Unreachable.
    return null;
  }

  @Override
  public Object visitBinaryExpr(Expr.Binary expr) {
    Object left = evaluate(expr.left);
    Object right = evaluate(expr.right);

    switch (expr.operator.type) {
    case BANG_EQUAL:
      return !isEqual(left, right);
    case EQUAL_EQUAL:
      return isEqual(left, right);
    case GREATER:
      checkNumberOperands(expr.operator, left, right);
      return (double)left > (double)right;
    case GREATER_EQUAL:
      checkNumberOperands(expr.operator, left, right);
      return (double)left >= (double)right;
    case LESS:
      checkNumberOperands(expr.operator, left, right);
      return (double)left < (double)right;
    case LESS_EQUAL:
      checkNumberOperands(expr.operator, left, right);
      return (double)left <= (double)right;
    case MINUS:
      checkNumberOperands(expr.operator, left, right);
      return (double)left - (double)right;
    case SLASH:
      checkNumberOperands(expr.operator, left, right);
      return (double)left / (double)right;
    case STAR:
      checkNumberOperands(expr.operator, left, right);
      return (double)left * (double)right;
    case PLUS:
      if (left instanceof Double && right instanceof Double) {
        return (double)left + (double)right;
      }

      if (left instanceof String && right instanceof String) {
        return (String)left + (String)right;
      }
      throw new RuntimeError(expr.operator,
                             "Operands must be two numbers or two strings.");
    }

    // Unreachable.
    return null;
  }

  @Override
  public Void visitExpressionStmt(Stmt.Expression stmt) {
    evaluate(stmt.expression);
    return null;
  }

  @Override
  public Void visitPrintStmt(Stmt.Print stmt) {
    Object value = evaluate(stmt.expression);
    System.out.println(stringify(value));
    return null;
  }

  @Override
  public Void visitVarStmt(Stmt.Var stmt) {
    Object value = null;
    if (stmt.initializer != null) {
      value = evaluate(stmt.initializer);
    }

    environment.define(stmt.name.lexeme, value);
    return null;
  }

  @Override
  public Object visitVariableExpr(Expr.Variable expr) {
    return environment.get(expr.name);
  }

  @Override
  public Object visitAssignExpr(Expr.Assign expr) {
    Object value = evaluate(expr.value);

    environment.assign(expr.name, value);
    return value;
  }

  @Override
  public Void visitBlockStmt(Stmt.Block stmt) {
    executeBlock(stmt.statements, new Environment(environment));
    return null;
  }

  void interpret(List<Stmt> statements) {
    try {
      for (Stmt statement : statements) {
        execute(statement);
      }
    } catch (RuntimeError error) {
      Lox.runtimeError(error);
    }
  }
}
